### PR TITLE
Add `no-for-loop-extra-variable` rule

### LINKED ISSUE #295: A case not handled by `no-for-loop`
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/295)

```js
/* eslint unicorn/no-for-loop: 2 */

const arr = [];

for (let i = 0, j = arr.length; i < j; i += 1) {
  const element = arr[i];
  console.log(element);
}
```
`no-for-loop` should report on this but it does not.

This slightly unusual pattern is used all over the place in `eslint-plugin-react`.

This issue is a subtask of #250.

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Sponsors (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by jmoore914**
Fixes #295

### CODE REVIEW

File: rules/no-for-loop-extra-variable.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+module.exports = {
+	create,
+	meta: {
+		type: 'suggestion',
+		docs: {
+			url: getDocumentationUrl(__filename)
+		},
+		fixable: 'code',
+		messages: {
+			noForLoopExtraVariable: 'Unnecessary variable in for loop.'
```
Comment (sindresorhus) [inline rules/no-for-loop-extra-variable.js]:
```suggestion
			noForLoopExtraVariable: 'Unnecessary variable in for-loop.'
```

File: test/no-for-loop-extra-variable.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+			errors: [error],
+			output: outdent`
+			if(conditional){
+				console.log(j)
+			}
+			else{	
+			var j = arr.length
+			for(var i = 1; j > i; i += 1) {
+					console.log(j);
+				}
```
Comment (sindresorhus) [inline test/no-for-loop-extra-variable.js]: Many of these examples have messy code-style, which makes them hard to read. Would be good if you could fix that.

File: docs/rules/no-for-loop-extra-variable.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Do not declare an extra variable in a for-loop setup
```
Comment (sindresorhus) [inline docs/rules/no-for-loop-extra-variable.md]: I don't think "setup" is the correct term here.

File: readme.md [hunk 4]
```diff
@@ -105,6 +106,7 @@ Configure it in `package.json`.
 - [no-console-spaces](docs/rules/no-console-spaces.md) - Do not use leading/trailing space between `console.log` parameters. *(fixable)*
 - [no-fn-reference-in-iterator](docs/rules/no-fn-reference-in-iterator.md) - Prevent passing a function reference directly to iterator methods. *(fixable)*
 - [no-for-loop](docs/rules/no-for-loop.md) - Do not use a `for` loop that can be replaced with a `for-of` loop. *(partly fixable)*
+- [no-for-loop-extra-variable](docs/rules/no-for-loop-extra-variable.md) - Do not declare an extra variable in a `for` loop setup *(fixable)*
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
- [no-for-loop-extra-variable](docs/rules/no-for-loop-extra-variable.md) - Do not declare an extra variable in a for-loop setup. *(fixable)*
```

File: docs/rules/no-for-loop-extra-variable.md (hunk 3 above)
Comment (fisker) [reply docs/rules/no-for-loop-extra-variable.md]: Maybe "init"

File: rules/no-for-loop-extra-variable.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
+'use strict';
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+function checkForLoop(context, node) {
+	if (checkTwoDeclaredVariables(node) && checkIteratorType(node)) {
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]: Pass `node.init` and `node.update` to those two functions.

File: rules/no-for-loop-extra-variable.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+function getIteratorName(node) {
+	return node.update.type === 'AssignmentExpression' ? node.update.left.name : node.update.argument.name;
+}
+
+function fixForLoop(node, context, fixer) {
+	const sourceCode = context.getSourceCode();
+	const nonIterator = getVariableNames(node).filter(variableName => variableName !== getIteratorName(node))[0];
+	const nonIteratorDeclaration = node.init.declarations.filter(declaration => declaration.id.name === nonIterator)[0];
+	const iteratorDeclaration = node.init.declarations.filter(declaration => declaration.id.name !== nonIterator)[0];
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]: Use `Array#find`

File: rules/no-for-loop-extra-variable.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+			return fixer.replaceText(token, nonIteratorInitValue);
+		});
+	}
+
+	function removeDeclarationFix() {
+		return [fixer.replaceText(node.init, node.init.kind + ' ' + sourceCode.getText(iteratorDeclaration))];
+	}
+
+	function moveDeclarationFix() {
+		return removeDeclarationFix().concat([fixer.insertTextBefore(node, node.init.kind + ' ' + sourceCode.getText(nonIteratorDeclaration) + '\n')]);
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]: Use `...`

File: rules/no-for-loop-extra-variable.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	return getNonIteratorTokenOutsideNode(sourceCode, node, nonIterator).length !== 0;
+}
+
+function getNonIteratorTokenOutsideNode(sourceCode, node, nonIterator) {
+	const tokens = node.init.kind === 'var' ? sourceCode.getTokensBefore(node) : [];
+	tokens.push(...sourceCode.getTokensAfter(node));
+	return tokens.filter(token => {
+		return token.type === 'Identifier' && token.value === nonIterator;
+	});
+}
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]:
```js
return [
...(node.init.kind === 'var' ? sourceCode.getTokensBefore(node) : []),
...sourceCode.getTokensAfter(node))
]
.filter(token => {
		return token.type === 'Identifier' && token.value === nonIterator;
	});
```

on cellphone. You got what I mean

File: rules/no-for-loop-extra-variable.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+			fix(fixer) {
+				return fixForLoop(node, context, fixer);
+			}
+		});
+	}
+}
+
+function checkTwoDeclaredVariables(init) {
+	return init.type === 'VariableDeclaration' && init.declarations.length === 2;
+}
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]:
```suggestion
function checkTwoDeclaredVariables(node) {
	return node.type === 'VariableDeclaration' && node.declarations.length === 2;
}
```

File: rules/no-for-loop-extra-variable.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+function checkIteratorType(node, update) {
+	let iteratorNode;
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]: Sorry, I didn't notice that node is used here, if so, maybe better pass `node` to this function

File: docs/rules/prefer-ternary.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
+# Prefer `.flatMap(…)` over `.map(…).flat()`
```
Comment (fisker) [inline docs/rules/prefer-ternary.md]: This file is not related

File: rules/no-for-loop-extra-variable.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+module.exports = {
+	create,
+	meta: {
+		type: 'suggestion',
+		docs: {
+			url: getDocumentationUrl(__filename)
+		},
+		fixable: 'code',
+		messages: {
+			noForLoopExtraVariable: 'Unnecessary variable in for-loop.'
```
Comment (fisker) [inline rules/no-for-loop-extra-variable.js]: Should use `message` for this case

```js
context.report({
	node,
	message,
	fix
})

File: docs/rules/no-for-loop-extra-variable.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
+# Do not declare an extra variable in a `for` loop init
```
Comment (snowteamer) [inline docs/rules/no-for-loop-extra-variable.md:1]: Nit: prefer the term "loop initializer" instead of "loop init".

File: readme.md [hunk 14]
```diff
@@ -102,6 +103,7 @@ Configure it in `package.json`.
 - [no-console-spaces](docs/rules/no-console-spaces.md) - Do not use leading/trailing space between `console.log` parameters. *(fixable)*
 - [no-fn-reference-in-iterator](docs/rules/no-fn-reference-in-iterator.md) - Prevent passing a function reference directly to iterator methods. *(fixable)*
 - [no-for-loop](docs/rules/no-for-loop.md) - Do not use a `for` loop that can be replaced with a `for-of` loop. *(partly fixable)*
+- [no-for-loop-extra-variable](docs/rules/no-for-loop-extra-variable.md) - Do not declare an extra variable in a for-loop init. *(fixable)*
```
Comment (snowteamer) [inline readme.md:106]: Same nit as above, prefer `initializer`

**Comment by sindresorhus** [comment]
Bump :)

File: docs/rules/no-for-loop-extra-variable.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+A pattern in some codebases is to declare a second variable in the initial state which is then only used in the test condition. This variable is unecessary and prevents the `no-for-loop` rule from flagging the loop.
+
+In the case where the second variable is used, either inside the loop, or is defined with `var` and used elsewhere in the code, it often makes more sense to define the variable outside the loop so that the intent of the loop is easier to understand.
```
[the remark this answers is not in the document]
Comment (jmoore914) [reply docs/rules/no-for-loop-extra-variable.md:5]: It wouldn't make sense to define the variable with `let` and use it in other places in the code because the scope of the variable would be contained to the for loop. If using `var` then the variable can be used in other places.

**Comment by sindresorhus** [comment]
@jmoore914 Is this ready for another round of review?

**Comment by jmoore914** [comment]
@sindresorhus it's still failing integration tests in a weird name (somehow nulls keep making it through when it shouldn't be possible to have a null in those spots) but if you want to start reviewing I can work on those changes since they won't change much

**Comment by sindresorhus** [comment]
@jmoore914 Are you still interested in finishing this? I need to know before we spend more time reviewing this.

**Comment by jmoore914** [comment]
@sindresorhus I don't have the time to finish this right now. If somebody else has the time, it's pretty close to done, otherwise I can come back to it another time.

**Comment by sindresorhus** [comment]
I understand. I'm going to close this for now to let give someone else a chance to finish this.
